api/diagnostics.go

```diff
@@ -298,118 +258,125 @@ func (j *DiagnosticsJob) flushReport(zipWriter *zip.Writer, fileName string, rep
 	}
 }
 
-func (j *DiagnosticsJob) collectDataFromNodes(ctx context.Context, nodes []dcos.Node, summaryReport *bytes.Buffer,
-	summaryErrorsReport *bytes.Buffer) ([]string, error) {
-
-	fetchRequests := j.getEndpointsToFetch(ctx, nodes, summaryReport, summaryErrorsReport)
-
-	fetchReq := make(chan fetcher.EndpointRequest, len(fetchRequests))
-	for _, r := range fetchRequests {
-		fetchReq <- r
-	}
-	close(fetchReq)
+func (j *DiagnosticsJob) downloadFromNodes(ctx context.Context, bundleID string,
+	nodes []dcos.Node, zipWriter *zip.Writer, summaryReport *bytes.Buffer, summaryErrorsReport *bytes.Buffer) {
 
-	fetchStatusUpdate := make(chan fetcher.StatusUpdate)
-	fetchResponse := make(chan fetcher.BulkResponse)
+	requests := j.getBundleRequests(bundleID, nodes, summaryReport)
+	requestFinished := make(chan fetcher.RequestStatus)
 
-	numberOfWorkers := j.Cfg.FlagDiagnosticsBundleFetchersCount
-	for i := 0; i < numberOfWorkers; i++ {
-		f, err := fetcher.New(j.Cfg.FlagDiagnosticsBundleDir, j.client, fetchReq, fetchStatusUpdate, fetchResponse, j.FetchPrometheusVector)
+	// send creation requests to all nodes
+	for _, r := range requests {
+		err := r.SendCreationRequest(ctx, requestFinished)
 		if err != nil {
-			return nil, fmt.Errorf("could not start fetchers: %s", err)
+			logrus.Errorf("error sending creation request to %s: %s", r.Node.IP, err)
 		}
-		go f.Run(ctx)
 	}
 
-	j.waitForStatusUpdates(ctx, fetchStatusUpdate, len(fetchRequests), summaryReport, summaryErrorsReport)
-
-	zips, errs := gatherAllResults(fetchResponse, numberOfWorkers)
-
-	if len(errs) != 0 {
-		j.logError(fmt.Errorf("%v", errs), "failed to gather all results", summaryErrorsReport)
+	// wait until all requests have reported that they're finished
+	for i := 0; i < len(requests); i++ {
+		select {
+		case <-ctx.Done():
+			return
+		case update := <-requestFinished:
```

Comment:
Here we can download the bundle directyle and send it to the zipQueue (i.e. call `downloadBundle`)